### PR TITLE
Konsti resnet implementation

### PR DESCRIPTION
**Pull request by KonstiNik**
Implementation of a flax ResNet from HuggingFace.
- Implement a pre-defined model. In that way, a pre-trained model can easily be fine-tuned.

### CODE REVIEW

**Comment by SamTov** [comment]
Thanks for the PR! In this case, I would have said the Black part should have been done down the chain in a separate PR. It makes it very difficult to review the larger changes to the code as there are now 109 files that need looking into. Can you highlight which modules you have changed in the ResNetPR? Alternatively, make a new PR to `main` where you only do the black formatting and then merge that one here.

File: examples/ResNet-Example.ipynb [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+   "execution_count": null,
+   "id": "dba15f7c",
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "data_generator = nl.data.CIFAR10Generator(2)\n",
+    "\n",
+    "# Input data needs to have shape (num_points, channels, height, width)\n",
```
Comment (SamTov) [inline examples/ResNet-Example.ipynb:46]: This is quite a funny problem. I also encountered it in some of our other projects. Is it worth setting this to the default? Just a question not something for this PR.

File: CI/unit_tests/models/test_huggingface_flax_model.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+--------
+If you use this module please cite us with:
+
+Summary
+-------
+"""
+
+import os
+
+os.environ["CUDA_VISIBLE_DEVICES"] = "-1"
```
Comment (SamTov) [inline CI/unit_tests/models/test_huggingface_flax_model.py:30]: Is this necessary? GitHub as no GPUs so it wont show up here and if you run it locally with a GPU it should theoretically work...

File: CI/unit_tests/models/test_huggingface_flax_model.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+        """
+        ntk = self.model.compute_ntk(self.x)["empirical"]
+        assert ntk.shape == (3, 3)
+
+    def test_infinite_failure(self):
+        """
+        Test that the call to the infinite NTK fails.
+        """
+        with pytest.raises(NotImplementedError):
+            self.model.compute_ntk(self.x, infinite=True)
```
Comment (SamTov) [inline CI/unit_tests/models/test_huggingface_flax_model.py:94]: I think there should be a newline here

File: znnl/models/jax_model.py [hunk 4]
```diff
@@ -58,27 +60,39 @@ def __init__(
...
                 optimizer to use in the training. OpTax is used by default and
                 cross-compatibility is not assured.
-        input_shape : tuple
-                Shape of the NN input.
+        input_shape : Optional[tuple]
+                Shape of the NN input. Required if no pre-built model is passed.
         seed : int, default None
                 Random seed for the RNG. Uses a random int if not specified.
-        ntk_batch_size : int, default 10
+        ntk_batch_size : Optional[int], default 10
```
Comment (SamTov) [inline znnl/models/jax_model.py:67]: Thanks for these updates. They are very helpful but also tedious to actually do, so I appreciate it.

**Review by SamTov — CHANGES_REQUESTED** [review]
I think this is one of the biggest (importance wise) PRs that ZnNL has seen for a long time, so awesome work. I have a couple of points scattered throughout the code but also two things I want to raise here:

1. Can we add one or two training pases to the test?
2. Did you not have to update the training procedure? The time I got this working, I needed to take into consideration the batch statistics and all these other things being passed correctly. I don't see these changes here, what was the solution?

File: examples/ResNet-Example.ipynb (hunk 1 above)
Comment (KonstiNik) [reply examples/ResNet-Example.ipynb:46]: Do you mean the data generator of 2 points, or the swapped axes of the input?

File: CI/unit_tests/models/test_huggingface_flax_model.py (hunk 2 above)
Comment (KonstiNik) [reply CI/unit_tests/models/test_huggingface_flax_model.py:30]: No it's not I will remove it

File: CI/unit_tests/models/test_huggingface_flax_model.py (hunk 3 above)
Comment (KonstiNik) [reply CI/unit_tests/models/test_huggingface_flax_model.py:94]: Will be added

File: examples/ResNet-Example.ipynb (hunk 1 above)
Comment (SamTov) [reply examples/ResNet-Example.ipynb:46]: The way they pass images with the channels in the first non-batch axis

File: CI/unit_tests/models/test_huggingface_flax_model.py (hunk 3 above)
Comment (SamTov) [reply CI/unit_tests/models/test_huggingface_flax_model.py:94]: I'm more concered as to why black didnt catch that

File: examples/ResNet-Example.ipynb (hunk 1 above)
Comment (KonstiNik) [reply examples/ResNet-Example.ipynb:46]: I found this very strange actually, as you still have to define the shape similar to as we currently do it. But the actual data is passed differently. 
We could theoretically adapt to that, but we could also implement e.g. a kwarg in the data generators to transform data to HF input shapes.
What do you think?

File: CI/unit_tests/models/test_huggingface_flax_model.py (hunk 3 above)
Comment (KonstiNik) [reply CI/unit_tests/models/test_huggingface_flax_model.py:94]: We should keep an eye on that.

**Comment by KonstiNik** [comment]
> 2. Did you not have to update the training procedure? The time I got this working, I needed to take into consideration the batch statistics and all these other things being passed correctly. I don't see these changes here, what was the solution?

No, I did not have to. The HF call method is directly compatible with constructing a Trainstate. After constructing it the rest is straight forward. Where exactly did you run into issues?

**Comment by SamTov** [comment]
> > 2. Did you not have to update the training procedure? The time I got this working, I needed to take into consideration the batch statistics and all these other things being passed correctly. I don't see these changes here, what was the solution?
> 
> No, I did not have to. The HF call method is directly compatible with constructing a Trainstate. After constructing it the rest is straight forward. Where exactly did you run into issues?

The call to the network has a new return signature? It should return the batch stats along with the logits and these batch stats have to be propagated to the network in the forward passes and during updates. We deal with this in the NTK calculation but unless it snuck through the last time I worked on it, there won't be batch stats passed

**Comment by KonstiNik** [comment]
> > > 2. Did you not have to update the training procedure? The time I got this working, I needed to take into consideration the batch statistics and all these other things being passed correctly. I don't see these changes here, what was the solution?
> > 
> > 
> > No, I did not have to. The HF call method is directly compatible with constructing a Trainstate. After constructing it the rest is straight forward. Where exactly did you run into issues?
> 
> The call to the network has a new return signature? It should return the batch stats along with the logits and these batch stats have to be propagated to the network in the forward passes and during updates. We deal with this in the NTK calculation but unless it snuck through the last time I worked on it, there won't be batch stats passed

Batch_stats are included in our model_state.params. But I have to check whether the batch_stats get handled properly.

**Comment by SamTov** [comment]
> > > > 2. Did you not have to update the training procedure? The time I got this working, I needed to take into consideration the batch statistics and all these other things being passed correctly. I don't see these changes here, what was the solution?
> > > 
> > > 
> > > No, I did not have to. The HF call method is directly compatible with constructing a Trainstate. After constructing it the rest is straight forward. Where exactly did you run into issues?
> > 
> > 
> > The call to the network has a new return signature? It should return the batch stats along with the logits and these batch stats have to be propagated to the network in the forward passes and during updates. We deal with this in the NTK calculation but unless it snuck through the last time I worked on it, there won't be batch stats passed
> 
> Batch_stats are included in our model_state.params. But good point, I have to check whether the batch_stats get handled properly.

[Here](https://github.com/google/flax/blob/main/examples/imagenet/train.py) for example, each time they call the model, they collect this other part of the output tuple:

```python
logits, new_model_state = state.apply_fn(
        {'params': params, 'batch_stats': state.batch_stats},
        batch['image'],
        mutable=['batch_stats'],
```
It isn't the batch stats sorry it is this model state part. This has to be passed to other functions in order for the model to train properly. From my memory, it had something to do with ensuring the batch stats are used and updated correctly. We don't do this in normal training we just ignore this additional output.

File: examples/ResNet-Example.ipynb (hunk 1 above)
Comment (SamTov) [reply examples/ResNet-Example.ipynb:46]: For me, our approach makes more sense because for each pixel we assign N numbers for colour. This approach would say that each colour get M pixels which I find funny. In some sense it is up to the user to know what they want their data to look like, but we do provide generators for this purpose. I would prefer to stick to a single shape for those generators though. I think the correct thing to do would be to stick to the Hugging face structure as this is the unchangeable thing. Our models thusfar are all defined at initialization so swapping an axis is easy.

**Comment by KonstiNik** [comment]
Further Changes need to include:
- [ ] Adapt Training Strategies to the updated TrainState 
- [ ] Jit larger training functions
- [ ] Adapt trace opt to updated TrainState

**Comment by SamTov** [comment]
> Further Changes need to include:
> 
> * [x]  Adapt Training Strategies to the updated TrainState
> * [ ]  Jit larger training functions
> * [ ]  Adapt trace opt to updated TrainState
> * [ ]  Make ResNet example more descriptive

You can remove traceopt from this, I can take care of it in my other traceopt PR. I have made enough changes to it there that this would just set the whole thing back.

**Comment by KonstiNik** [comment]
There was an issue with HF FlaxResNets when using smaller models with layer_type='basic' instead of layer_type='bottleneck':
https://github.com/huggingface/transformers/issues/27257
It is fixed and merged into the main branch of hf-transformers. So it might be released soon. 

For my examples and tests to pass, I have therefore used layer_type='bottleneck'.

File: znnl/training_strategies/training_steps.py [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+from typing import Callable
+
+import jax.random
+from flax.training.train_state import TrainState
+from jax import jit, lax
+
+logger = logging.getLogger(__name__)
+
+
+class TrainStep:
```
Comment (SamTov) [inline znnl/training_strategies/training_steps.py:39]: Can you add some doc strings here? I think we should also have a test for this class as far as it is possible. It is called in the `training_strategies` test but it's always better to unit test stuff like this and make sure nothing is broken.

File: znnl/training_strategies/training_decorator.py [hunk 6]
```diff
@@ -92,6 +92,17 @@ def wrapper(trainer: "SimpleTraining", *args, **kwargs):
         Output of the train_fn. A more detailed documentation can be found in the used
         training strategy.
         """
+        # Check whether a model is available
+        if trainer.model is None:
+            raise ValueError("No model available. Please initialize a model first.")
+        # Check whether the model has been set up properly
+        if trainer._train_step is None:
+            raise ValueError(
+                "The model has not been set up properly."
```
Comment (SamTov) [inline znnl/training_strategies/training_decorator.py:101]: This exception isn't very helpful from my perspective. If I were reading this after a model failed I wouldn't really know what to call to fix it. Can you add more information like` call model.set_model()` or something more direct.

File: znnl/models/jax_model.py [hunk 7]
```diff
@@ -58,34 +83,52 @@ def __init__(
...
         self.input_shape = input_shape
 
         # Initialized in self.init_model
         self.rng = None
 
+        # Input shape is required if no full model is passed.
+        if pre_built_model is None and input_shape is None:
+            raise ValueError(
+                "Input shape must be specified if no pre-built model is passed."
+                "Model is yet to be constructed."
```
Comment (SamTov) [inline znnl/models/jax_model.py:115]: What do you think about `model has not been constructed` to keep it all in simple English.

File: znnl/models/huggingface_flax_model.py [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+        params : dict
+                Network parameters to use in the calculation.
+        x : np.ndarray
+                Data on which to apply the network
+
+        Returns
+        -------
+        Acts on the data with the model architecture and parameter set.
+        """
+        # out = self.apply_fn(
```
Comment (SamTov) [inline znnl/models/huggingface_flax_model.py:113]: Can you please remove the commented out code?

File: znnl/models/flax_model.py [hunk 9]
```diff
@@ -165,18 +153,69 @@ def _init_params(self, kernel_init: Callable = None, bias_init: Callable = None)
...
+        Parameters
+        ----------
+        params: dict
+                Contains the model parameters to use for the model computation.
+                It is a dictionary of structure
+                {'params': params, 'batch_stats': batch_stats}
+        inputs : np.ndarray
+                Feature vector on which to apply the model.
+
+        TODO(Konsti): Make the apply function work with the batch_stats.
```
Comment (SamTov) [inline znnl/models/flax_model.py:169]: What is the issue with it?

File: znnl/models/flax_model.py [hunk 10]
```diff
@@ -100,6 +101,9 @@ def __init__(
                 The default value is trace_axes(-1,), which reduces the NTK to a tensor
                 of rank 2.
                 For a full NTK set trace_axes=().
+        store_on_device : bool, default True
```
Comment (SamTov) [inline znnl/models/flax_model.py:104]: This is a cool feature, is it new in NT?

File: CI/unit_tests/models/test_huggingface_flax_model.py (hunk 2 above)
Comment (SamTov) [reply CI/unit_tests/models/test_huggingface_flax_model.py:30]: Just bumping this in case it was missed.

File: CI/integration_tests/models/test_huggingface_flax_model_deployment.py [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+--------
+If you use this module please cite us with:
+
+Summary
+-------
+"""
+
+import os
+
+os.environ["CUDA_VISIBLE_DEVICES"] = "-1"
```
Comment (SamTov) [inline CI/integration_tests/models/test_huggingface_flax_model_deployment.py:30]: Can we remove this one?

**Review by SamTov — CHANGES_REQUESTED** [review]
Such a cool PR! I am excited to see it in action. I did have a few more comments that it would be great if we could discuss.

File: znnl/models/flax_model.py (hunk 9 above)
Comment (KonstiNik) [reply znnl/models/flax_model.py:169]: Right now and also before the PR, there is no proper handling of batch statistics in the flax class. If you would want to have batch normalization within a model it would not work properly.
I want to fix that (in a different PR) to allow flax models to have batchstats.

File: znnl/models/flax_model.py (hunk 10 above)
Comment (KonstiNik) [reply znnl/models/flax_model.py:104]: I actually don't know. have seen it for the first time too and put it in.

File: znnl/training_strategies/training_steps.py (hunk 5 above)
Comment (KonstiNik) [reply znnl/training_strategies/training_steps.py:39]: Done in [0fd55b7](https://github.com/zincware/ZnNL/pull/105/commits/0fd55b7fe1343d50d2ae0d4bfaedeeb98359096e) and [9b8021e](https://github.com/zincware/ZnNL/pull/105/commits/9b8021e72812c84f815f5723a84833320976c533)

File: znnl/training_strategies/training_decorator.py (hunk 6 above)
Comment (KonstiNik) [reply znnl/training_strategies/training_decorator.py]: Fixed in [6f75030](https://github.com/zincware/ZnNL/pull/105/commits/6f75030536a93d9c5a2279f433b9fb00e756d5e2)

File: znnl/models/jax_model.py (hunk 7 above)
Comment (KonstiNik) [reply znnl/models/jax_model.py]: Fixed in [a87d2e6](https://github.com/zincware/ZnNL/pull/105/commits/a87d2e6bbc614d1986483630d85799c38a57ee47)

File: znnl/models/huggingface_flax_model.py (hunk 8 above)
Comment (KonstiNik) [reply znnl/models/huggingface_flax_model.py]: Done in [339a436](https://github.com/zincware/ZnNL/pull/105/commits/339a436bb58d2c09694cb6db2aeecc47887f1cdf)

File: CI/integration_tests/models/test_huggingface_flax_model_deployment.py (hunk 11 above)
Comment (KonstiNik) [reply CI/integration_tests/models/test_huggingface_flax_model_deployment.py]: Removed in [e4b04f5](https://github.com/zincware/ZnNL/pull/105/commits/e4b04f5a4b2da4b89e14d2f5fde28b499c962d76)

**Review by SamTov — APPROVED** [review]
Noice
